train_decoder.py

```diff
@@ -0,0 +1,462 @@
+    trainer = trainer.to(device=inference_device)
+    
+    send_to_device = lambda arr: [x.to(device=inference_device, dtype=torch.float) for x in arr]
+    step = start_step
+    for epoch in range(start_epoch, epochs):
+        print(f"=========== Starting epoch {epoch} ===========")
+        decoder.train()
+        sample = 0
+        last_sample = 0
+        last_snapshot = 0
+        last_time = time.time()
+        losses = []
+        for i, (img, emb) in enumerate(dataloaders["train"]):
+            step += 1
+            sample += img.shape[0]
+            img, emb = send_to_device((img, emb))
+            
+            for unet in range(1, trainer.num_unets+1):
+                loss = trainer.forward(img, image_embed=emb, unet_number=unet)
+                loss.backward()
+                trainer.update(unet_number=unet)
+                losses.append(loss.item())
+
+            samples_per_sec = (sample - last_sample) / (time.time() - last_time)
+            last_time = time.time()
+            last_sample = sample
+
+            if i % 10 == 0:
+                average_loss = sum(losses) / len(losses)
+                print(f"Epoch {epoch}/{epochs} Sample {sample} Step {i} - {samples_per_sec:.2f} samples/sec")
+                print(f"Losses: {losses}")
+                print(f"Loss: {average_loss}")
+                print("")
+                if using_wandb:
+                    wandb.log({
+                        "Training loss": average_loss,
+                        "Epoch": epoch,
+                        "Sample": sample,
+                        "Step": i,
+                        "Samples per second": samples_per_sec
+                    }, step=step)
+                losses = []
+
+            if last_snapshot + save_every_n_samples < sample:
+                last_snapshot = sample
+                print(f"Saving model...")
+                save_trainer(base_path, trainer, epoch, step, validation_losses, local_only=not using_wandb, latest=save_latest, best=False, checkpoint=save_all)
+                if using_wandb:
+                    print(f"Generating sample...")
+                    train_images = generate_samples(decoder, dataloaders["train"], epoch, inference_device, step, n=n_sample_images, text_prepend="Train: ")
+                    wandb.log({
+                        "Train samples": train_images
+                    }, step=step)
+
+            if epoch_samples is not None and sample >= epoch_samples:
+                break
+
+        print(f"=========== Starting Validation {epoch} ===========")
+        with torch.no_grad():
+            decoder.eval()
+            sample = 0
+            average_loss = 0
+            start_time = time.time()
+            for i, (img, emb, txt) in enumerate(dataloaders["val"]):
+                sample += img.shape[0]
+                img, emb = send_to_device((img, emb))
+                
+                with autocast(enabled=amp):
```

Comment:
the trainer class should take care of this

Reply:
oh actually, nevermind, i don't handle this just yet, let me think about it

Reply:
ok, now it should handle it, as long as you place the trainer into eval mode (then call forward)